### PR TITLE
Beethoven strategy

### PR DESCRIPTION
**Pull request by olegnat**

### CODE REVIEW

File: contracts/base/strategies/beethoven/BeethovenBase.sol [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    pool = _pool;
+    poolId = _poolId;
+    beethovenVault = IBeethovenVault(_beethovenVault);
+    depositToken = _depositToken;
+    fundManagementStruct = IBeethovenVault.FundManagement(address(this), false, payable(address(this)), false);
+    beethovenPoolId = _beethovenPoolId;
+    rewardToDepositPoolId = _rewardToDepositPoolId;
+    setupPullTokens();
+
+    //    ISpookyMasterChef.PoolInfo memory poolInfo = ISpookyMasterChef(pool).poolInfo(_poolID);
```
Comment (belbix) [inline contracts/base/strategies/beethoven/BeethovenBase.sol:82]: remove please

File: contracts/base/strategies/beethoven/BeethovenBase.sol [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  function rewardToUnderlying(address rewardToken, uint toCompound) internal{
+    if (toCompound == 0) {
+      return;
+    }
+    if (rewardToken != depositToken) {
+      balancerSwap(rewardToDepositPoolId, rewardToken, depositToken, toCompound);
+    }
+    uint depositTokenBalance = IERC20(depositToken).balanceOf(address (this));
+    IERC20(depositToken).safeApprove(address(beethovenVault), 0);
+    IERC20(depositToken).safeApprove(address(beethovenVault), depositTokenBalance);
```
Comment (belbix) [inline contracts/base/strategies/beethoven/BeethovenBase.sol:190]: better to move approves inside `balancerJoin`

File: test/CoreContractsWrapper.ts [hunk 3]
```diff
@@ -33,6 +33,8 @@ export class CoreContractsWrapper {
   public announcerLogic: string;
   public vaultController: VaultController;
   public vaultControllerLogic: string;
+  public newCalculator: PriceCalculator | null = null;
```
Comment (belbix) [inline test/CoreContractsWrapper.ts:36]: nope, calculator not a core contract

File: test/strategies/UniversalStrategyTest.ts [hunk 4]
```diff
@@ -74,13 +75,23 @@ async function universalStrategyTest(
...
       deployInfo.vault = vault;
       deployInfo.strategy = strategy;
-
+      let calculator =  deployInfo?.tools?.calculator as PriceCalculator;
+      // temporary solution until new Price calculator deployed
+      const isBTP = await StrategyTestUtils.isBalancerLP(user, underlying);
+      if (isBTP && deployInfo?.core?.controller != null && deployInfo.tools?.calculator != null) {
+        const newCalculator = (await DeployerUtils.deployPriceCalculator(signer, deployInfo.core.controller.address))[0];
+        for (const bptLp of FtmAddresses.BPT_LPs) {
+          await newCalculator.registerBPT(bptLp);
```
Comment (belbix) [inline test/strategies/UniversalStrategyTest.ts:84]: move registration to deploy function

File: contracts/base/strategies/beethoven/BeethovenBase.sol (hunk 1 above)
Comment (olegnat) [reply contracts/base/strategies/beethoven/BeethovenBase.sol:82]: done.

File: contracts/base/strategies/beethoven/BeethovenBase.sol (hunk 2 above)
Comment (olegnat) [reply contracts/base/strategies/beethoven/BeethovenBase.sol:190]: done

File: test/CoreContractsWrapper.ts (hunk 3 above)
Comment (olegnat) [reply test/CoreContractsWrapper.ts]: fixed

File: test/strategies/UniversalStrategyTest.ts (hunk 4 above)
Comment (olegnat) [reply test/strategies/UniversalStrategyTest.ts]: refactored this logic. Not needed anymore.

File: contracts/base/strategies/beethoven/BeethovenBase.sol [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+/// @title Abstract contract for Beethoven strategy implementation
+/// @author OlegN
+abstract contract BeethovenBase is StrategyBase {
+  using SafeMath for uint256;
+  using SafeERC20 for IERC20;
+
+  // ************ VARIABLES **********************
+  /// @notice Strategy type for statistical purposes
+  string public constant override STRATEGY_NAME = "BeethovenStrategyACBase";
```
Comment (belbix) [inline contracts/base/strategies/beethoven/BeethovenBase.sol:28]: use contract name please

Comment (olegnat) [reply contracts/base/strategies/beethoven/BeethovenBase.sol]: updated

File: contracts/base/strategies/beethoven/BeethovenBase.sol [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+* as all warranties, including any fitness for a particular purpose with respect
+* to Tetu and/or the underlying software and the use thereof are disclaimed.
+*/
+
+pragma solidity 0.8.4;
+
+import "../StrategyBase.sol";
+import "../../../third_party/beethoven/IBeethovenxChef.sol";
+import "../../../third_party/beethoven/IBeethovenVault.sol";
+import "hardhat/console.sol";
```
Comment (belbix) [inline contracts/base/strategies/beethoven/BeethovenBase.sol:18]: remove please

File: contracts/strategies/fantom/beethoven/StrategyBeethoven.sol [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+* or implied. Any use of this open source software released under the ISC
+* Internet Systems Consortium license is done at your own risk to the fullest
+* extent permissible pursuant to applicable law any and all liability as well
+* as all warranties, including any fitness for a particular purpose with respect
+* to Tetu and/or the underlying software and the use thereof are disclaimed.
+*/
+pragma solidity 0.8.4;
+
+
+import "../../../third_party/uniswap/IUniswapV2Pair.sol";
```
Comment (belbix) [inline contracts/strategies/fantom/beethoven/StrategyBeethoven.sol:15]: remove

File: contracts/infrastructure/price/PriceCalculator.sol [hunk 8]
```diff
@@ -45,6 +47,7 @@ contract PriceCalculator is Initializable, Controllable, IPriceCalculator {
   uint256 constant public DEPTH = 20;
   address public constant CRV_USD_BTC_ETH_MATIC = 0xdAD97F7713Ae9437fa9249920eC8507e5FbB23d3;
   address public constant CRV_USD_BTC_ETH_FANTOM = 0x58e57cA18B7A47112b877E31929798Cd3D703b0f;
+  address public constant BEETHOVEN_VAULT_FANTOM = 0x20dd72Ed959b6147912C2e529F0a0C651c33c9ce;
```
Comment (belbix) [inline contracts/infrastructure/price/PriceCalculator.sol:50]: increase version of the contract

File: test/ToolsContractsWrapper.ts [hunk 9]
```diff
@@ -22,7 +22,7 @@ export class ToolsContractsWrapper {
   public readonly zapContract: ZapContract;
   public readonly multicall: Multicall;
   public readonly pawnshopReader: PawnShopReader;
-
+  public newCalculator: PriceCalculator | null = null;
```
Comment (belbix) [inline test/ToolsContractsWrapper.ts]: remove

Comment (olegnat) [reply test/ToolsContractsWrapper.ts]: done

File: contracts/infrastructure/price/PriceCalculator.sol (hunk 8 above)
Comment (olegnat) [reply contracts/infrastructure/price/PriceCalculator.sol:50]: done

File: contracts/strategies/fantom/beethoven/StrategyBeethoven.sol (hunk 7 above)
Comment (olegnat) [reply contracts/strategies/fantom/beethoven/StrategyBeethoven.sol:15]: done

File: contracts/base/strategies/beethoven/BeethovenBase.sol (hunk 6 above)
Comment (olegnat) [reply contracts/base/strategies/beethoven/BeethovenBase.sol:18]: done

**Pull request merged**